plugin-server/src/worker/vm/promise-manager.ts

```diff
@@ -0,0 +1,25 @@
+import { PluginsServerConfig } from '../../types'
+
+export class PromiseManager {
+    pendingPromises: Set<Promise<any>>
+    config: PluginsServerConfig
+
+    constructor(config: PluginsServerConfig) {
+        this.pendingPromises = new Set()
+        this.config = config
+    }
+
+    public trackPromise(promise: Promise<any>): void {
+        this.pendingPromises.add(promise)
+
+        promise.finally(() => {
+            this.pendingPromises.delete(promise)
+        })
+    }
+
+    public async awaitPromisesIfNeeded(): Promise<void> {
+        while (this.pendingPromises.size > this.config.MAX_PENDING_PROMISES_PER_WORKER) {
+            await Promise.race(this.pendingPromises)
```

Comment:
https://github.com/PostHog/posthog/pull/8573#issuecomment-1041345850

I think `Promise.race` is actually even better suited here. 

cc @mariusandra